azure-iot-device/tests/common/test_http_transport.py

```diff
@@ -0,0 +1,279 @@
+            fake_client_cert.pass_phrase,
+        )
+
+
+class HTTPTransportTestConfig(object):
+    @pytest.fixture
+    def mock_http_client_constructor(self, mocker):
+        mocker.patch.object(ssl, "SSLContext").return_value
+        mocker.patch.object(HTTPTransport, "_create_ssl_context").return_value
+        mock_client_constructor = mocker.patch.object(http_client, "HTTPSConnection", autospec=True)
+        mock_client = mock_client_constructor.return_value
+        response_value = mock_client.getresponse.return_value
+        response_value.status = 1234
+        response_value.reason = "__fake_reason__"
+        response_value.read.return_value = "__fake_response_read_value__"
+        return mock_client_constructor
+
+    @pytest.fixture
+    def transport(self, mock_http_client_constructor):
+        return HTTPTransport(hostname=fake_hostname)
+
+
+@pytest.mark.describe("HTTPTransport - .request()")
+class TestRequest(HTTPTransportTestConfig):
+    @pytest.mark.it("Generates a unique HTTP Client connection for each request")
+    def test_creates_http_connection_object(self, mocker, transport, mock_http_client_constructor):
+        # We call .result because we need to block for the Future to complete before moving on.
+        transport.request(fake_method, fake_path, mocker.MagicMock()).result()
+        assert mock_http_client_constructor.call_count == 1
+
+        transport.request(fake_method, fake_path, mocker.MagicMock()).result()
+        assert mock_http_client_constructor.call_count == 2
+
+    @pytest.mark.it("Uses the HTTP Transport SSL Context.")
+    def test_uses_ssl_context(self, mocker, transport, mock_http_client_constructor):
+        ssl_context_mocker = transport._ssl_context
+        done = transport.request(fake_method, fake_path, mocker.MagicMock())
+        done.result()
+
+        assert mock_http_client_constructor.call_count == 1
+        assert mock_http_client_constructor.call_args[1]["context"] == ssl_context_mocker
+
+    @pytest.mark.it("Formats the request URL correctly given a path and no query parameters.")
+    def test_formats_http_client_request_with_only_method_and_path(
+        self, mocker, transport, mock_http_client_constructor
+    ):
+        mock_http_client_request = mock_http_client_constructor.return_value.request
+        fake_method = "__fake_method__"
+        fake_path = "__fake_path__"
+        expected_url = "https://{}/{}".format(fake_hostname, fake_path)
+        done = transport.request(fake_method, fake_path, mocker.MagicMock())
+        done.result()
+
+        assert mock_http_client_constructor.call_count == 1
+        assert mock_http_client_request.call_count == 1
+        assert mock_http_client_request.call_args[0][0] == fake_method
+        assert mock_http_client_request.call_args[0][1] == expected_url
+        assert mock_http_client_request.call_args[1]["body"] == ""
+        assert bool(mock_http_client_request.call_args[1]["headers"]) is False
```

Comment:
Here and everywhere - try to avoid checking individual arguments whenever possible